Added role whitelist to internals of wp-cli guest author generation command.

smistephen opened the pull request:
Previously, as noted in https://github.com/Automattic/Co-Authors-Plus/issues/553, `wp co-authors-plus create-guest-authors` created a guest author for every user in the database.

This was a massive performance suck, as well as being largely unnecessary. Adding a subscriber as a guest author is not a common use case.

Now, only users having the role of administrator, editor, and author have guest authors generated for them.

The query to filter users by role is a touch heavy (it performs LIKEs on meta values), but testing it on a site with 1000 users showed no major issues.

It was also compared to simply querying all the users, and doing the filtering in PHP. Head to head, there was no significant difference in performance between the two approaches, and using the WordPress API is always a better bet than writing your own one-off custom filter algorithm.

Commit:
CoAuthors display now for Pages and JS is being enqueued on Pages

Commit:
JS error goes away from js_vars()

Commit:
changed is_post_type_enabled() to clean up code

Commit:
removed extra whitespace

Commit:
removed all traces of $post_type in functions, since is_post_type_enabled() should do the trick now

trepmal committed:
Revert "Update version constant"

This reverts commit 525f211c38b56e6b52bbb1e8b88e8f408dc411be.

Commit:
Edited words to match glossary

Commit:
Removed continue because the script would stop running.

Commit:
Added better consistencies with glossary

trepmal committed:
Merge branch 'master' into reduce-sleep

trepmal committed:
Revert "Revert "Update version constant""

This reverts commit 5ac7d9a2288d49db6462a98d9211dfa42480a0f7.

Commit:
doesn't return false after quick edit saves

Commit:
replaced hardcoded 'author' with $this->coauthor_taxonomy

philipjohn committed:
Tested tag to 4.8

philipjohn committed:
Merge pull request #400 from Automattic/reduce-sleep

Reduce amount of sleep

philipjohn committed:
Merge pull request #420 from rebeccahum/cli-create-terms-for-posts

Fixed WP CLI create-terms-for-posts if no co-authors found

philipjohn committed:
Merge pull request #423 from rebeccahum/terminology

Terminology updated throughout

philipjohn committed:
Merge pull request #426 from rebeccahum/replace_hardcoded_author

Replace hardcoded 'author' with $this->$coauthor_taxonomy

trepmal committed:
Move parenthesis to fix esc_html and sprintf

Fixes warning: `Warning: sprintf(): Too few arguments in /path/wp-content/plugins/co-authors-plus/php/class-coauthors-guest-authors.php on line 487`

Warning surfaces when deleting a guest author that is mapped to a WP user

trepmal committed:
Move parenthesis to fix esc_html and sprintf

Fixes warning: `Warning: sprintf(): Too few arguments in /path/wp-content/plugins/co-authors-plus/php/class-coauthors-guest-authors.php on line 487`

Warning surfaces when deleting a guest author that is mapped to a WP user

trepmal committed:
Move parenthesis to fix esc_html and sprintf

Fixes warning: `Warning: sprintf(): Too few arguments in /path/wp-content/plugins/co-authors-plus/php/class-coauthors-guest-authors.php on line 487`

Warning surfaces when deleting a guest author that is mapped to a WP user

Commit:
Added progress so users have an idea of how long it will take

philipjohn committed:
Merge pull request #430 from Automattic/trepmal-patch-1

Move parenthesis to fix esc_html and sprintf

Commit:
Added a better style of progress bar!

Commit:
Deleting guest authors is less confusing

Commit:
Fixed merge conflicts

Commit:
Merge remote-tracking branch 'refs/remotes/Automattic/master'

Commit:
Hopefully now merge conflict is gone now...

Commit:
some extra notes.

Commit:
Merge remote-tracking branch 'refs/remotes/Automattic/master' into delete_guest_author

# Conflicts:
#	php/class-coauthors-guest-authors.php

Commit:
featured image to avatar

Commit:
featured image to avatar

Commit:
featured image to avatar

Commit:
Removed avatar sizing

Commit:
Removed avatar sizing

Commit:
Removed avatar sizing

Commit:
remove duplicated linked author byline

Commit:
remove duplicated linked author byline

Commit:
remove duplicated linked author byline

Commit:
coauthors_wp_list_authors has the option now of listing only guest authors

philipjohn committed:
Merge pull request #433 from rebeccahum/featured_image_avatar

Guest author's featured image is avatar now

philipjohn committed:
Merge pull request #434 from rebeccahum/remove_avatar_sizing

Removed extra image sizing

philipjohn committed:
Merge pull request #435 from rebeccahum/remove_duplicated_byline

Remove duplicated byline

Commit:
remove duplicates from linked accounts

Commit:
changes as per philipjoin

Commit:
Added later escaping

Commit:
Merge remote-tracking branch 'refs/remotes/Automattic/master'

Commit:
Use linked account for accurate post count

Commit:
Redid README.md with better markdown formatting

Commit:
Redid README.md with better markdown formatting

Commit:
added readme.txt back

Commit:
added readme.txt back

Commit:
remove quotes, stray =s and upgrade notice section

Commit:
remove quotes, stray =s and upgrade notice section

Commit:
added check for false $term

trepmal committed:
Do not 404 when guest author archive has 0 posts. Addresses #425 (#440)

Resolves #425, #409

trepmal committed:
Do not 404 when guest author archive has 0 posts. Addresses #425 (#440)

Resolves #425, #409

Commit:
Merge remote-tracking branch 'refs/remotes/Automattic/master'

Commit:
Merge branch 'master' of https://github.com/rebeccahum/Co-Authors-Plus

Commit:
Merge remote-tracking branch 'refs/remotes/Automattic/master'

Commit:
fixed author archive title

Commit:
revised comment

jordie23 committed:
Fix invalid CSS (#442)

jordie23 committed:
Fix invalid CSS (#442)

Commit:
fixed coauthors_link_single to work with guest authors

Commit:
added _blank target

Commit:
switched logic around

Commit:
Merge remote-tracking branch 'refs/remotes/Automattic/master'

Commit:
added hooks for creating/deleting guest authors

Commit:
added hooks for creating/deleting guest authors

sboisvert committed:
Merge pull request #446 from rebeccahum/added_guest-author_hooks

Added guest author hooks for create/delete

The hooks are useful in being able to manage caches that take into account guest authors.

Commit:
Merge remote-tracking branch 'refs/remotes/Automattic/master'

Commit:
Merge pull request #439 from rebeccahum/new_readme.MD

New README.md

Commit:
Merge remote-tracking branch 'refs/remotes/Automattic/master'

sboisvert committed:
Merge pull request #438 from rebeccahum/accurate_guest_post_count

Accurate Guest Author post count on linked accounts

Commit:
fix logic for doing_autosave check

Commit:
Merge remote-tracking branch 'refs/remotes/Automattic/master'

Commit:
Merge branch 'master' of https://github.com/rebeccahum/Co-Authors-Plus

Commit:
Merge remote-tracking branch 'refs/remotes/origin/user_login-spaces' into upstream/master

Commit:
Fixes "Notice: Trying to get property of non-object" for users who have user_login with spaces

Commit:
Fixes issue of not being able to reassign posts to user with user_login with spaces on deletion of a user

philipjohn committed:
Merge pull request #437 from rebeccahum/dedupe_coauthors_wp_list_authors

remove duplicates from linked accounts on coauthors_wp_list_authors()

philipjohn committed:
Merge pull request #436 from rebeccahum/coauthors_wp_list_authors

coauthors_wp_list_authors() has option to list only guest authors now

philipjohn committed:
Merge pull request #432 from rebeccahum/delete_guest_author

Deleting guest authors is less confusing

philipjohn committed:
Merge pull request #407 from haleeben/fix/no_authors_found_message

Author search UX issues

philipjohn committed:
Merge pull request #441 from rebeccahum/filter_author_archive

Filter author archive

philipjohn committed:
Merge pull request #414 from mdchiragpatel/fixed-120

Remove associated guest user when mapped user id deleted.

philipjohn committed:
Merge pull request #453 from rebeccahum/user_login_spaces

user_login spaces problem when using add_coauthors

Commit:
Adding escaping

Commit:
Merge remote-tracking branch 'refs/remotes/Automattic/master'

pkevan committed:
Adding details of filter for slow performance

- Use simple tax queries for CAP to improve performance
- Detail of filter to add to theme
- Adding CLI command to run

philipjohn committed:
Merge pull request #444 from rebeccahum/fix_coauthors_links_single

Fix coauthors_links_single()

philipjohn committed:
Merge pull request #456 from Automattic/fix/issue319

Adding details of filter for slow performance

philipjohn committed:
Remove redundant test for 404 on Author Archive

Since we made CAP consistent with core in it's handling of author archives where there are no posts (see https://github.com/Automattic/Co-Authors-Plus/issues/425) we no longer expect there to be a 404 response on the author archive for a guest author.

Commit:
Merge remote-tracking branch 'refs/remotes/Automattic/master'

Commit:
Fixed counting for guest author posts

Commit:
Users table will give accurate post count if linked guest author has post(s)

Commit:
No need to add $guest_count->count anymore now that the Users table displays the count properly

Commit:
Added one more validation

sboisvert committed:
Merge pull request #461 from rebeccahum/guest_author_count_table

Guest Author Counts are more accurate

sboisvert committed:
Merge pull request #450 from mslinnea/fix-autosave-logic

Fixes logic for DOING_AUTOSAVE check

sboisvert committed:
Merge pull request #422 from rebeccahum/pages_quick_edit

Pages archive now displays coauthors and quick edit works

sboisvert committed:
Merge pull request #386 from trepmal/milestone/kl-375

Fix private post viewing on front-end

rebeccahum committed:
Merge pull request #431 from rebeccahum/wpcli_create-guest-author_progress

Added progress to create-guest-authors so users have an idea of how long it will take

Commit:
Merge remote-tracking branch 'refs/remotes/Automattic/master'

Commit:
set $coauthors_loading

rebeccahum committed:
Merge pull request #468 from rebeccahum/set-coauthors_loading

Set $coauthors_loading

sanketio committed:
[#198] Fix guest authors with non-ASCII characters

sanketio committed:
[#198] Replace .html() with .text() for simple string in js

sanketio committed:
[#279] Fix duplicate name display issue

 - When `coauthors_auto_apply_template_tags` is set to true, `coauthors_posts_links()` or `the_author_posts_link()` methods displays duplicate names if post has multiple authors.

sanketio committed:
[#198] Improve unit testing for non-ASCII character issue

sanketio committed:
[#279] Add issue link as ref

sanketio committed:
[#279] Improve phpunit test cases

sanketio committed:
[#198] Improve phpunit test cases + cosmetic changes

sanketio committed:
[#198] Assign values to variables instead of keeping empty for test cases

sanketio committed:
[#279] Improve condition checking for unit test cases

sanketio committed:
[#279] Refactor comments for test cases

sanketio committed:
[#184] Started unit tests for template tags

- Added unit tests for get_coauthors() & is_coauthor_for_post()

sanketio committed:
[#184] Unit tests for coauthors() for template tags

sanketio committed:
[#184] Unit tests for coauthors_posts_links_single() for template tags

sanketio committed:
[#184] Unit tests for coauthors_firstnames() for template tags

sanketio committed:
[#184] Method description for coauthors_firstnames() for template tags

sanketio committed:
[#184] Unit tests for coauthors_lastnames() for template tags

sanketio committed:
[#184] Unit tests for coauthors_nicknames() for template tags

sanketio committed:
[#184] Unit tests for coauthors_emails() for template tags

sanketio committed:
[#184] Unit tests for coauthors_links_single() for template tags

sanketio committed:
[#184] Unit tests for coauthors_ids() for template tags

sanketio committed:
[#184] Unit tests for get_the_coauthor_meta() for template tags

sanketio committed:
[#184] Unit tests for coauthors_wp_list_authors() for template tags

sanketio committed:
[#184] Unit tests for coauthors_get_avatar() for template tags

sanketio committed:
[#184] Fix undefined variable warnings and add feed_type for feed link

sanketio committed:
[#388] Unit tests for is_guest_authors_enabled() method

sanketio committed:
[#388] Unit tests for get_coauthor_by() method

sanketio committed:
[#184] Improve unit tests

sanketio committed:
[#388] Unit tests for is_post_type_enabled() method

sanketio committed:
[#388] Fix unit tests for is_post_type_enabled() method using current screen

sanketio committed:
[#184] Remove not needed assertions and add proper commenting

sanketio committed:
[#388] Unit tests for current_user_can_set_authors() method

- Also fixed warning "Trying to get property of non-object" if post_type is not set in "get_current_screen()" on https://github.com/Automattic/Co-Authors-Plus/blob/master/co-authors-plus.php#L1014

sanketio committed:
[#388] Unit tests for search_authors() method

sanketio committed:
[#388] Unit tests for get_author_term() method

sanketio committed:
[#388] Unit tests for update_author_term() method

sanketio committed:
[#388] Unit tests for get_guest_author_by() method

sanketio committed:
[#388] Unit tests for get_guest_author_thumbnail() method

sanketio committed:
[#388] Unit tests for get_guest_author_fields() method

sanketio committed:
[#388] Unit tests for get_all_linked_accounts() method

sanketio committed:
[#388] Unit tests for create_guest_author_from_user_id() method

sanketio committed:
[#388] Unit tests for handle_delete_guest_author_action() method

sanketio committed:
[#388] Unit tests for CoAuthors_Guest_Authors::delete() method

sanketio committed:
[#388] Unit tests for coauthors orders

sanketio committed:
[#388] Improve unit tests

sanketio committed:
[#184] Unit tests for coauthors orders

sanketio committed:
[#388] Remove unit tests for coauthors orders

sanketio committed:
[#388] Improve unit tests

megfh committed:
Update class-wp-cli.php

fix #472

rebeccahum committed:
Merge pull request #480 from megfh/patch-1

Update class-wp-cli.php

sanketio committed:
[#198] Replace urlencode/urldecode with rawurlencode/rawurldecode

sanketio committed:
[#279] Code cleanup

- Convert duplicate conditions into variables and use those variables.

sanketio committed:
[#388] Modify where condition when WordPress generates `post_author IN (id)` in query

philipjohn committed:
Merge pull request #473 from sanketio/fix/198

Fix the issue where guest authors with non-ASCII characters can't be used as co-authors

sanketio committed:
[#458] Update travis file

nmeh20 committed:
Changes to resolve issue #332 about missing coauthor meta in get_the_coauthor_meta function

nmeh20 committed:
Changes to resolve issue #332. Updated funtion resolve error generated while fetching 'type' field.

nmeh20 committed:
Changes for optimizing code of get_the_coauthor_meta function

sanketio committed:
[#458] Include WP 4.6 env in travis file

philipjohn committed:
Merge pull request #482 from sanketio/fix/458

Update .travis.yml file for PHPUnit tests

philipjohn committed:
Merge pull request #476 from sanketio/fix/388

Review and improve test coverage

philipjohn committed:
Merge pull request #475 from sanketio/fix/184

Unit tests/Fix warnings for template tags

philipjohn committed:
Merge pull request #457 from Automattic/fix/author_404_test

Remove redundant test for 404 on Author Archive

nmeh20 committed:
Updating function to allow fetching meta data based on user_id to support the_coauthor_meta function

nmeh20 committed:
Removing unused global variables

sanketio committed:
Merge branch 'master' into fix/279

# Conflicts:
#	tests/test-template-tags.php

nmeh20 committed:
Formatting in my changes as per PHPCS

nmeh20 committed:
Space alignment and adding escape to output

philipjohn committed:
Merge pull request #474 from sanketio/fix/279

Fix the issue where incompatibility when `coauthors_auto_apply_template_tags` set to true

philipjohn committed:
Release v3.2: Update version numbers etc

philipjohn committed:
Merge pull request #484 from nilzari/master

Changes to resolve issue #332 about missing coauthor meta

philipjohn committed:
Release v3.3: Changelog and props

sanketio committed:
Fix failing test cases

philipjohn committed:
Merge pull request #500 from sanketio/fix/499

Fix failing test cases for coauthors functions

philipjohn committed:
Merge branch 'release/3.3'

binodkalathil committed:
Unexpected output of template tag 'coauthors_links' is resolved by updating conditions

The template tag was outputting usernames of guest authors when they do not have a website set in their profile. It will now output guest authors display names as expected.

binodkalathil committed:
Corrected PHPUnit test cases to work with the udpated code

A couple of test cases were failing as the return value of a function has changed as per the modifications done for fixing the issue(Issue #469)

pkevan committed:
Bump last version of WP to 4.8

binodkalathil committed:
Merge branch 'master' into issues/469-function-return-author-login-name-instead-of-display-name

sboisvert committed:
Merge pull request #504 from binodkalathil/issues/469-function-return-author-login-name-instead-of-display-name

Unexpected output of template tag 'coauthors_links' is resolved by updating conditions

rebeccahum committed:
Revert "Merge pull request #438 from rebeccahum/accurate_guest_post_count"

This reverts commit 2953a4cc26ab9d1d46ad25b726b355ab00e39943, reversing
changes made to 32c7f09e4aac73c9b59962c21f3a02b4640e96f4.

blunce24 committed:
Added i18n support to Author title and name for author archive

blunce24 committed:
changed author display name on author archive page to be Full Name rather than slug

blunce24 committed:
added check for  before using it in query

blunce24 committed:
updated truthiness check for current_coauthor_term object

blunce24 committed:
formatting adjustments

blunce24 committed:
formatting updates

blunce24 committed:
allow PHPCS inspection of file

TheCrowned committed:
Replacing search_fields argument (non-existent) in get_users() call with search_columns.

rebeccahum committed:
Merge pull request #516 from blunce24/i18n-author-archive

Added i18n support and corrected author display name for author archive: addressing issue #515

blunce24 committed:
fixed tab issue

blunce24 committed:
changed check on coauthor term to be more concise

sboisvert committed:
Merge pull request #517 from blunce24/current-user-no-terms-warning

check for coauthor terms before using them in query: addressing issue #511

philipjohn committed:
Merge pull request #519 from TheCrowned/fix/513-author-search-argument

Replacing search_fields with search_columns in get_users()

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Add isset() checkes to post registration

Commit:
remove trailing whitespace

TheCrowned committed:
Updated readme with FAQ on how to disable Guest Authors. Fix #122.

mdbitz committed:
Merge pull request #521 from paulschreiber/fix/warnings

Adds isset() checks to post type registration

TheCrowned committed:
Fix test_coauthors_links_single_when_guets_author(), which was wrongly passing.

The test passed because was updating a postmeta instead of the user record in wp_users, thus subsequent assertions all passed simply because no real edit had been done.
When making the edit for real, the test would fail in several places, so tweaks were needed to fix its behavior.

TheCrowned committed:
Updated coauthors_links_single so that CAP coauthor link matches Core author link structure.

TheCrowned committed:
Removed function action_pre_user_query and its hooks, as it was obsolete

TheCrowned committed:
Fix possible discrepancy in author search due to mismatched use of user_login and user_nicename.

sboisvert committed:
Merge pull request #529 from TheCrowned/fix/505-test-wrong-meta-key

Test test_coauthors_links_singe_guests_authors wrongly passing

sboisvert committed:
Merge pull request #523 from TheCrowned/fix/122-faq-disable-guest-author

Added FAQ to readme on how to disable Guest Authors

jacobarriola committed:
Only filter author template for title

Repro steps:
- in `archive.php`, use `the_archive_title()` method
- load a category archive page

Expected results:
"Category: category_name"

Actual results:
null

Notes:
The `filter_author_archive_title()` function is hooked to the `get_the_archive_title` filter, which can run in other archive templates such as `category` or `tag`. When run in such a template, the function does not return the `$title` parameter, resulting in `null`.

sboisvert committed:
Merge pull request #531 from TheCrowned/fix/530-remove-action_pre_user_query

Removed function action_pre_user_query and its hooks

TheCrowned committed:
Removed whole setUp function, which contained the two filters that made WP queries non-TEMPORARY

TheCrowned committed:
Skipping irrelevant cap checks through list of allowed caps.

TheCrowned committed:
Merge branch 'master' of https://github.com/Automattic/Co-Authors-Plus into fix/489-skip-filtering-for-irrelevant-caps

mariovalney committed:
Allow guest authors have personal data exported (#528)

* [Style] Identation changed to tab

* [Feat] Added personal data exporter to guest authors

* [Feat] Added a filter to allow third part plugins add guest author data

* [Style] WPCS fixes

* [Fix] Filter name changed

* [Style] WPCS fixes

* [Fix] Filter name changed

TheCrowned committed:
Updated tests to search_authors() to use assertArrayNotHasKey instead of assertNotContains

TheCrowned committed:
Re-enabled Test_Author_Queries->test__author_arg__user_is_coauthor__author_arg()

TheCrowned committed:
Re-enabled Test_Author_Queries->tests__author_name_arg_plus_tax_query__is_coauthor()

TheCrowned committed:
Improved inline doc

sboisvert committed:
Merge pull request #546 from TheCrowned/fix/538-re-enable-tests

Re-enable disabled Test_Author_Queries tests

Commit:
Add post type parameter to the Mine link, so it works for Pages

rebeccahum committed:
Merge pull request #547 from pdemarte/post-type-pages-fix

Add post type parameter to the Mine link, so it works for Pages

jacobarriola committed:
Merge branch 'master' into patch-1

rebeccahum committed:
Merge pull request #535 from jacobarriola/patch-1

🐛Only filter author template for title

rebeccahum committed:
Merge pull request #532 from TheCrowned/fix/273-search-discrepancy

Fix possible discrepancy in author search with ignored authors

rebeccahum committed:
Merge pull request #543 from TheCrowned/fix/489-skip-filtering-for-irrelevant-caps

Skip irrelevant capabilities filtering in filter_user_has_cap() with whitelist

TheCrowned committed:
Changed get_user_by('slug' to get_user_by('login' when using linked_account

mdbitz committed:
Merge pull request #540 from TheCrowned/fix/526-non-temporary-queries-on-multisite-test

Removed Test_Author_Queried_Object setUp function

mdbitz committed:
Merge pull request #558 from TheCrowned/fix/554-user-login-linked-account

column_posts() should use user_login when querying for linked_account

TheCrowned committed:
Only using new `cap-` prefixed term when counting user posts.

sboisvert committed:
Merge pull request #567 from TheCrowned/fix/561-wrong-post-count-guest-authors

Fixing double posts counts for users with linked accounts

TheCrowned committed:
$ignored_authors in search should use user_nicename, not login.

rebeccahum committed:
Merge pull request #569 from TheCrowned/fix/568-failing-test

$ignored_authors in search should use user_nicename, not login.

smistephen committed:
Added role whitelist to internals of wp-cli guest author generation command.

Previously, as noted in https://github.com/Automattic/Co-Authors-Plus/issues/553, `wp co-authors-plus create-guest-authors` created a guest author for every user in the database.

This was a massive performance suck, as well as being largely unnecessary. Adding a subscriber as a guest author is not a common use case.

Now, only users having the role of administrator, editor, and author have guest authors generated for them.

The query to filter users by role is a touch heavy (it performs LIKEs on meta values), but testing it on a site with 1000 users showed no major issues.

It was also compared to simply querying all the users, and doing the filtering in PHP. Head to head, there was no significant difference in performance between the two approaches, and using the WordPress API is always a better bet than writing your own one-off custom filter algorithm.

TheCrowned commented on php/class-wp-cli.php:
```diff
@@ -25,7 +25,7 @@ public function create_guest_authors( $args, $assoc_args ) {
 		$this->args = wp_parse_args( $assoc_args, $defaults );
 
-		$users = get_users();
+		$users = get_users( array( 'role__in' => array( 'administrator', 'editor', 'author' ) ) );
```
I think we may want to allow filtering of this in some way. Some sites may have other roles they want to be allowed to be added as coauthors. Since it's a CLI command we may want to allow an optional argument to overwrite the default list of roles :)

Also, I think the `contributor` role should be included by default, as it has the `edit_posts` cap and thus permission to be added as coauthor :)

smistephen replied:
Works for me. After lunch I'll put together some thoughts on a --roles flag :)

smistephen committed:
Add cli option to allow users to set custom role whitelist.

The option defaults to creating guest authors for administrators, editors, authors, and contributors. Roles are provided in a comma separated list such as `--roles=author,freelancer`.

If the user provides a role that does not exist in their instance, execution is halted and an error is displayed such as `Role freelancer does not exist.`. This error message is made available for translation.

smistephen committed:
Reset pot file permissions to 755 from 644

smistephen committed:
Ban subscribers from being made guest authors via wp-cli.

Some sites can have millions of subscribers. Querying for that many users presents a serious potential performance challenge.

Rather than take the risk of having an end user specify them in the --roles flag, they have now been explicitly banned.

smistephen committed:
Modify user filtering algorithm.

Filtering users by role on the database has the potential to make for a very expensive query, as it performs a LIKE on an unindexed field.

But if the users were all brought over at once, unfiltered, the potential exists for an array with millions of elements, straining the application server's memory limit.

So the compromise is: batching.

Users are fetched from the database, unfiltered, in batches of user-defined size. They are then filtered, and a new batch is brought in.

This algorithm looks to strike a balance between load on the database server, and memory limits on the application server.

sboisvert commented on php/class-wp-cli.php:
```diff
@@ -21,11 +48,77 @@ public function create_guest_authors( $args, $assoc_args ) {
+			// Subscribers cannot be made guest authors en masse
+			// due to potential performance issues (sites can have millions of subscribers).
+			if ( 'subscriber' === $role ) {
+				WP_CLI::error( __( 'For performance reasons, subscribers cannot be made guest authors via wp-cli. Please use the admin interface.', 'co-authors-plus' ) );
```
Maybe we should have a "force" option?

smistephen replied:
Sure, I can add that

sboisvert commented on php/class-wp-cli.php:
```diff
@@ -21,11 +48,77 @@ public function create_guest_authors( $args, $assoc_args ) {
-				// There are no arguments at this time
-			);
+			'roles' => array( 'administrator', 'editor', 'author', 'contributor' ),
+			'batch-size' => 1000,
```
Why would folks need to specify this? What would be the difference if we were to limit it to batches of 100 at a time?

smistephen replied:
It's to stop the script from fetching **ALL** of the users at once and crashing servers with low memory limits, while also allowing servers with greater resources to open up the throttle.

smistephen replied:
If we were to limit it to batches of 100 at a time, it would take longer but use less memory (a tradeoff I've had to make many times in the past running on low-end equipment, haha)

sboisvert commented on php/class-wp-cli.php:
```diff
@@ -21,11 +48,77 @@ public function create_guest_authors( $args, $assoc_args ) {
+		foreach ( $role_whitelist as $role ) {
+			// Does this role exist in this instance?
+			if ( ! $GLOBALS['wp_roles']->is_role( $role ) ) {
+				WP_CLI::error( __( 'Role ', 'co-authors-plus' ) . $role . __( ' does not exist.', 'co-authors-plus' ) );
```
Should we instead inject $role as a variable inside a singular translation? That would seem to be best for many languages where the sentence structure may be different.

smistephen replied:
:+1:

sboisvert commented on php/class-wp-cli.php:
```diff
@@ -21,11 +48,77 @@ public function create_guest_authors( $args, $assoc_args ) {
+			}
+		}
+
+		$users_unfiltered_query = new WP_User_Query( array( 'number' => $batch_size ) );
```
Do we get any benefits if we order by primary key?

smistephen replied:
Not really, we're going to iterate over the entire array no matter what, so the order isn't important :)

sboisvert replied:
If we order by primary key, would it make the lookups and offsetting be faster?

smistephen replied:
Turns out yes, yes it would: https://github.com/smistephen/Co-Authors-Plus/commit/7231625d3d42c3b3680c3ba036baa430944a4f64

sboisvert reviewed (changes requested):
I felt a few comments inline.
There doesn't seem to be a dry run option and there doesn't seem to be a way to capture the log of what was changed. Could we add those?

smistephen committed:
Change method of translating invalid role error message.

Previously, the translated string was broken up into two fragments, with the $role variable concatenated in between.

To enable easier translation, the string is now generated in one piece via sprintf (interpolated variables inside double quotes are not allowed).

smistephen committed:
Add --force-subscribers flag to enable users to mass-generate subscriber guest authors.

In the event that an end user accepts the performance hit of generating guest authors for all their subscribers, this flag allows them to do so.

Named --force-subscribers instead of --force to allow for future development.

smistephen committed:
Add --not-a-dry-run flag controlling whether or not the command performs data modifications.

By default, the command outputs the IDs of the users it would have attempted to generate guest authors for, had this not been a dry run.

Passing in the --not-a-dry-run flag tells the code to actually make the attempts.

Props sboisvert for the idea to make this command dry by default.

smistephen committed:
Add --log-output flag to toggle output of guest author creation attempt results.

This flag results in one of two potential log messages:

"User ID 1 is now linked to Guest Author 2"
"Error while attempting to generate guest author for user ID 1: errormessagehere"

The command logs to STDOUT by default, but can be easily redirected to the location of the end user's choice.

TheCrowned commented on php/class-wp-cli.php:
```diff
@@ -21,28 +67,149 @@ public function create_guest_authors( $args, $assoc_args ) {
+			$role_whitelist = $this->args['roles'];
+		} else {
+			$role_whitelist = explode( ',', $this->args['roles'] );
+		}
```
This is really just minor and can be ignored, but I still believe worth pointing out. Cleaner:
```
$role_whitelist = $this->args['roles'];
if ( false === is_array( $role_whitelist ) ) {
	$role_whitelist = explode( ',', $role_whitelist );
}
```

smistephen replied:
Done :+1:

TheCrowned commented:
Hey! We may want to improve the inline documentation of the function to comply with WP_CLI PHPDoc standards. Resources that have been very helpful for me when I dealt with similar issue have been this [Commands Cookbook - Annotating with PHPDoc](https://make.wordpress.org/cli/handbook/commands-cookbook/#annotating-with-phpdoc) and [How to write custom WP_CLI commands for WordPress automation](https://firxworx.com/blog/wordpress/how-to-write-custom-wp-cli-commands-for-wordpress-automation/) (scroll to _PHPDoc comments and controls_ section). 

For example, we'd want to split the description in short and long; all optional args should be wrapped in `<` `>`; add a `synopsis`; etc.

Also, boolean options shouldn't be treated as all others. For example, specifying `[--<dry-run>]` will automatically force it to be a boolean arg, and will allow to disable with `--no-dry-run`. This way, you should also manage to get rid of all the `is_bool()` checks, and any boolean option can be turned off with a `no-` in front of it :) This way everything is simpler and we avoid having a `not-a-dry-run` => false` by default, which is a bit confusing.

On a different point, I don't think that the dry run mode should be the default one - let developers toggle it if they want :) (that is also as other CLI commands work)

Finally, if I understand @sboisvert intentions correctly, he meant we _should_ log the routine progress, but not necessarily give an option to do so. The logging should happen by default. Then, if someone doesn't want it, they can redirect it to `/dev/null`. Adding an option for that would really just be adding another way to write `> /dev/null`, and I don't think it's necesary :)

(I also left inline comments :))

smistephen commented on php/class-wp-cli.php:
```diff
@@ -21,28 +67,149 @@ public function create_guest_authors( $args, $assoc_args ) {
+				if ( is_wp_error( $result ) ) {
+					$skipped++;
+					if ( true === $log_output ) {
+						foreach ( $result->get_error_messages() as $error_message ) {
```
This is just me programming defensively.

`WP_Error->get_error_messages()` (https://developer.wordpress.org/reference/classes/wp_error/) returns an array of error messages even if there's just one.

I could use `WP_Error->get_error_message()`, which returns an int or string, but that makes me incredibly nervous, as it silently suppresses all other error messages even if they exist.

So yes, currently `create_guest_author_from_user_id()` only ever returns one error message. But I can't predict that it always will, so I coded it this way :)

TheCrowned replied:
Yes, it is possible. I was skeptical just because I have never seen that in any plugin/core piece I ever looked at, but yes :)

smistephen committed:
Change method of validating inputs

smistephen committed:
Add reference to new --force-subscribers flag in error message

smistephen committed:
Modifying documentation in accordance with the standard

smistephen committed:
Modify name of --not-a-dry-run flag to --dry-run for greater clarity.

The flag is now named --dry-run to take advantage of built-in command negation functionality, allowing users to pass --no-dry-run to turn it off.

smistephen committed:
Change sort column for user query.

Despite there being an index on the user_login column in wp_users, MySQL does not use it and instead performs a filesort.

So now the query orders by ID.

Props sboisvert for the subtle poke that something wasn't quite right with the query.

smistephen committed:
Spacing modifications in keeping with coding standards

smistephen commented on php/class-wp-cli.php:
```diff
@@ -21,28 +67,149 @@ public function create_guest_authors( $args, $assoc_args ) {
+			// Subscribers cannot be made guest authors en masse
+			// due to potential performance issues (sites can have millions of subscribers).
+			if ( 'subscriber' === $role && false === $force_subscribers ) {
+				WP_CLI::error( __( 'For performance reasons, subscribers cannot be made guest authors via wp-cli. Please use the admin interface.', 'co-authors-plus' ) );
```
Done :+1:

smistephen commented on php/class-wp-cli.php:
```diff
@@ -21,28 +67,149 @@ public function create_guest_authors( $args, $assoc_args ) {
+			$batch_size = (int) $this->args['batch-size'];
+		} else {
+			WP_CLI::error( __( 'Batch size must be an integer greater than zero.', 'co-authors-plus' ) );
+		}
```
Done :+1:

TheCrowned commented on php/class-wp-cli.php:
```diff
@@ -11,38 +11,225 @@
+			$force_subscribers = true;
+		} else {
+			$force_subscribers = false;
+		}
```
At this point we can just have a `$force_subscribers = $this->args['force_subscribers'];` , and the same applies to the other bool options :)

TheCrowned reviewed (commented):
One final thing: I'm not sure we want to include the command expected output in the example lines? I think it's enough to have a (brief) description of what it does, and then the command, without the related output :)

smistephen commented:
The output is in keeping with documentation specification.

smistephen committed:
Condense assignment operations for cli flags.

Due to changes in how the flags are read, we can now condense our checks into a series of oneliners.

GaryJones commented:
The key changes are going to need to be manually picked out of this, but there's value to be had. Adding to a future milestone.
